### PR TITLE
[SYCL][CUDA] Fixes context release and unnamed context scope

### PR DESCRIPTION
**Pull request by steffenlarsen**
This PR fixes the following bugs:

1. Some times upon deletion, the deleted context would stay on the CUDA context stack and cause subsequent CUDA operations to fail. To fix this the context stack is now checked after destruction and if the dead context is still there it is removed.
1. In one location a `ScopedContext` was created without being assigned to a variable and would therefore die immediately. It should now live for the entirety of the surrounding scope.

Signed-off-by: Steffen Larsen <steffen.larsen@codeplay.com>

### CODE REVIEW

File: sycl/plugins/cuda/pi_cuda.cpp [hunk 1]
```diff
@@ -1165,11 +1165,14 @@ pi_result cuda_piContextRelease(pi_context ctxt) {
     CUcontext cuCtxt = ctxt->get();
     CUcontext current = nullptr;
     cuCtxGetCurrent(&current);
-    if(cuCtxt != current)
-    {
-      PI_CHECK_ERROR(cuCtxSetCurrent(cuCtxt));
+    if (cuCtxt != current) {
```
Comment (romanovvlad) [inline sycl/plugins/cuda/pi_cuda.cpp]: Please, explain what the problem is in the commits message and add a test.

Comment (steffenlarsen) [reply sycl/plugins/cuda/pi_cuda.cpp]: Description has been added to both the commit and this PR. The bugs that this fixes are some odd corner cases and are difficult to test.

Comment (steffenlarsen) [reply sycl/plugins/cuda/pi_cuda.cpp]: Give me a moment. I pushed something to the wrong branch.
